### PR TITLE
fluids: SmartSim Online Data-Driven SGS training

### PR DESCRIPTION
**Pull request by jrwrigh**
Todo:
- [x] Documentation
- [x] Send subgrid stresses (not just filtered velocity products)
    - [x] Add SGS to dd training input DM
    - [x] Add calculation of subgrid stress to training input calculation
    - [x] Verify DM Vec data placement (if true, remove IS and copying mechanics)
- [x] Add test

Depends on #1223

### CODE REVIEW

**Comment by jrwrigh** [comment]
I currently don't have a test setup for this case. I could do one to test just the generation of the training data (so nothing about the communication with the database), but the current build framework I have requires building with SmartRedis for that. I think requiring building with SmartRedis makes sense, but not sure if we want to make a work around for that just for testing purposes.

File: examples/fluids/README.md [hunk 1]
```diff
@@ -8,6 +8,8 @@ The state variables are mass density, momentum density, and energy density.
 
 The main Navier-Stokes solver for libCEED is defined in [`navierstokes.c`](navierstokes.c) with different problem definitions according to the application of interest.
 
+## Build Instructions
```
Comment (jeremylt) [inline examples/fluids/README.md:11]: Minor
```suggestion
## Build instructions
```

File: examples/fluids/README.md [hunk 2]
```diff
@@ -231,7 +239,7 @@ For the case of a square/cubic mesh, the list of face indices to be used with `-
 
 ### Boundary conditions
 
-Boundary conditions for compressible viscous flows are notoriously tricky. Here we offer some recommendations
+Boundary conditions for compressible viscous flows are notoriously tricky. Here we offer some recommendations.
```
Comment (jeremylt) [inline examples/fluids/README.md:242]:
```suggestion
Boundary conditions for compressible viscous flows are notoriously tricky.
Here we offer some recommendations.
```

File: examples/fluids/problems/sgs_dd_training.c [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+    PetscCall(DMGetLocalVector(user->diff_filter->dm_filter, &FilteredFields_loc));
+
+    PetscCall(DifferentialFilterApply(user, solution_time, Q, FilteredFields));
+    PetscCall(DMGlobalToLocal(user->diff_filter->dm_filter, FilteredFields, INSERT_VALUES, FilteredFields_loc));
+
+    PetscCall(DMGetGlobalVector(sgs_dd_train->filtered_grad_velo_proj->dm, &FilteredVelocityGradient));
+    PetscCall(VelocityGradientProjectionApply(sgs_dd_train->filtered_grad_velo_proj, FilteredFields_loc, FilteredVelocityGradient));
+
+    // Compute Training Data
+    {
```
Comment (jrwrigh) [inline examples/fluids/problems/sgs_dd_training.c:298]: Note to self: Fix formatting

**Comment by jrwrigh** [comment]
I hereby declare this PR officially good for review!

**Comment by jrwrigh** [comment]
Alright, I've reviewed over this and rebased onto main. Good to review. @jedbrown @jeremylt

File: examples/fluids/include/smartsim.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+// This file is part of CEED:  http://github.com/ceed
+
+#include <c_client.h>
+#include <petscsys.h>
+#include <sr_enums.h>
+
+#if defined(__clang_analyzer__)
+void SmartRedisCall(SRError);
+#else
+#define SmartRedisCall(...)                                                                                                           \
```
Comment (jeremylt) [inline examples/fluids/include/smartsim.h:15]: For what its worth, I think PETSc would call this `PetscSmartRedisCall`

File: tests/junit_common.py [hunk 5]
```diff
@@ -312,6 +312,47 @@ def diff_cgns(test_cgns: Path, true_cgns: Path, tolerance: float = 1e-12) -> str
     return proc.stderr.decode('utf-8') + proc.stdout.decode('utf-8')
 
 
+def test_case_output_string(test_case: TestCase, spec: TestSpec, mode: RunMode,
```
Comment (jeremylt) [inline tests/junit_common.py:315]: This is a pretty big change to the common JUNIT code - have you tested that this core file can drop into Ratel with these changes? @zatkins-dev comments?

Comment (zatkins-dev) [reply tests/junit_common.py:315]: I'll review/test this afternoon with Ratel

Comment (zatkins-dev) [reply tests/junit_common.py:315]: This looks like a pretty clean refactor of the output string generation, I can't imagine it causing any issues. I'll double check but tentatively looks good

Comment (zatkins-dev) [reply tests/junit_common.py:315]: LGTM! Ratel tests ran without issue locally. If this is finalized, I can make a matching MR in Ratel for the changes.

**Comment by jrwrigh** [comment]
Thanks! I've changed the `SmartRedisCall` to `PetscSmartRedisCall` as Jeremy suggested. I've also rebased onto `main` just to make sure everything is working correctly CI-wise.

**Pull request merged**